garak/generators/azure.py

```diff
@@ -0,0 +1,140 @@
+# some azure openai model names should be mapped to openai names
+chat_models_openai = {
+    "gpt-4": "gpt-4-turbo-2024-04-09", 
+    "gpt-4o": "gpt-4o", # links to latest version
+    "gpt-4o-mini": "gpt-4o-mini", # links to latest version
+    "gpt-35-turbo": "gpt-3.5-turbo-0125", 
+    "gpt-35-turbo-16k": "gpt-3.5-turbo-16k", 
+    "gpt-4-32k": "gpt-4-32k", 
+}
+
+completion_models_openai = {
+    "gpt-35-turbo-instruct": "gpt-3.5-turbo-instruct", 
+    "davinci-002": "davinci-002", 
+    "babbage-002": "babbage-002", 
+}
+
+class AzureOpenAIChat(OpenAICompatible):
+    """Wrapper for Azure Open AI. Expects AZURE_API_KEY, AZURE_ENDPOINT and AZURE_DEPLOYMENT_NAME environment variables.
+
+    Uses the [OpenAI-compatible API](https://learn.microsoft.com/en-us/azure/ai-services/openai/api-version-deprecation)
+    via direct HTTP request.
+
+    To get started with this generator:
+    #. Visit [https://learn.microsoft.com/en-us/azure/ai-services/openai/concepts/models](https://learn.microsoft.com/en-us/azure/ai-services/openai/concepts/models) and find the LLM you'd like to use.
+    #. [Deploy a model](https://learn.microsoft.com/en-us/azure/ai-services/openai/how-to/create-resource?pivots=web-portal#deploy-a-model) and copy paste the model and deployment names
+    #. On the Azure portal page for the Azure OpenAI you want to use click on "Resource Management -> Keys and Endpoint" and copy paste the API Key and endpoint. 
+    #. In your console, Set the ``AZURE_API_KEY``, ``AZURE_ENDPOINT`` and ``AZURE_DEPLOYMENT_NAME`` variables.
+    #. Run garak, setting ``--model_type`` to ``azure`` and ``--model_name`` to [the name of the model](https://learn.microsoft.com/en-us/azure/ai-services/openai/concepts/models)
+    - e.g. ``gpt-4o``.
+    """
+
+    ENV_VAR = "AZURE_API_KEY"
+    active = True
+    generator_family_name = "Azure"
+    name_backup = ""
+
+    def __init__(self, name="", config_root=_config):
+        if name in chat_models_openai:
+            self.name = chat_models_openai[name]
+        elif name in completion_models_openai:
+            self.name = completion_models_openai[name]
+        else:
+            self.name = name
+
+        self._load_config(config_root)
+        if self.name in context_lengths:
+            self.context_len = context_lengths[self.name]
+
+        super().__init__(
+            self.name, generations=self.generations, config_root=config_root
+        )
+
+    def _load_client(self):
+        azure_endpoint = os.getenv("AZURE_ENDPOINT", default=None)
+        if azure_endpoint is None:
+            raise ValueError(
+                'The AZURE_ENDPOINT environment variable is required.\n'
+            )
+
+        azure_deployment_name = os.getenv("AZURE_DEPLOYMENT_NAME", default=None)
+        if azure_deployment_name is None:
+            raise ValueError(
+                'The AZURE_DEPLOYMENT_NAME environment variable is required.\n'
+            )
```

Comment:
Items from env variables are better enforced with override of `_validate_env_var()` as the values should not need to be evaluated again on subsequent calls to `_load_client()` that may occur. Also a pattern to allow injection of the environment variable to utilized via a configuration files is desired. Something like `endpoint_env_var` and `deployment_env_var` can pair are with constants in the the class to allow for this.

``` Python
DEPLOYMENT_ENV_VAR = "AZURE_DEPLOYMENT_NAME"
ENDPOINT_ENV_VAR = "AZURE_ENDPOINT"
```

See example: https://github.com/leondz/garak/blob/c6abfabe05a809617c5f726a8c6ab2b9436a27ac/garak/generators/nemo.py#L60-L73

```suggestion

def _validate_env_var(self):
        if self.azure_endpoint is None:
            if not hasattr(self, "endpoint_env_var"):
                self.endpoint_env_var = self.ENDPOINT_ENV_VAR
            self.azure_endpoint = os.getenv(self.endpoint_env_var, None)

            raise ValueError(
                f"The {self.endpoint_env_var} environment variable is required."
            )

        if self.azure_deploymnet_name is None:
            if not hasattr(self, "deployment_env_var"):
                self.deployment_env_var = self.DEPLOYMENT_ENV_VAR
            self.azure_deploymnet_name = os.getenv(self.endpoint_env_var, None)

            raise ValueError(
                f"The {self.endpoint_env_var} environment variable is required."
            )

        super()._validate_env_var()


    def _load_client(self):            
```

Reply:
should be fixed in the last commit